specs/marketplace.md

```diff
@@ -0,0 +1,374 @@
+
+### Filling slot
+
+When new Request is created `StorageRequested(requestId, ask, expiry)` event is emitted with following properties: 
+
+ - `requestId` - ID of Request.
+ - `ask` - Specification of Request parameters. For details see above.
+ - `expiry` - Unix timestamp that specifies when the Request will be cancelled if all slots are not filled by then.
+
+It is then up to the Storage Provider node to decide based on the emitted parameters if it wants to participate in the 
+Request and try to fill its slot(s). This decision SHOULD be done based on parameters specified by the node operator.
+If the node decides to ignore this Request, no action is necessary. If the node wants to try to fill a a slot, then 
+it MUST follow the remaining steps.
+
+The Node MUST decide which Slot specified by slot's index it wants to try to fill in. The Node MAY try filling multiple
+slots. In order to fill a slot, the node MUST first download the slot's data using the CID of the manifest (**TODO: Manifest RFC**) and the index of  the slot. The CID is  specified in `request.content.cid`, which can be retrieved from the smart contract using `getRequest(requestId)`.
+Then the node MUST generate a proof over the downloaded data (**TODO: Proving RFC**).
+
+When the proof is ready it then MUST create a transaction for smart contract call `fillSlot()` with following REQUIRED:
+ 
+ - Parameters:
+   - `requestId` - ID of the Request.
+   - `slotIndex` - Index that the node is trying to fill.
+   - `proof` - `Groth16Proof` proof structure, generated over the slot's data.
+ - The Ethereum address of the node from which the transaction originates MUST have [approval](https://docs.openzeppelin.com/contracts/2.x/api/token/erc20#IERC20-approve-address-uint256-) for transfer of at least the amount required as collateral for the Request on the ERC20 based token, that the network utilizes.
+
+If the proof is invalid, or the slot was already filled by another node, then the transaction
+will revert, otherwise a `SlotFilled(requestId, slotIndex)` event is emitted. If the transaction is successful, then the
+node SHOULD transition into a __proving__ state as it will need to submit proof of data possession when prompted by the
+contract.
+
+It should be noted that if the node sees the `SlotFilled` emitted for slot that he is downloading the dataset or 
+generating proof for, then node SHOULD stop and choose different non-filled slot to try to fill as the chosen slot
+was filled by other node.
+
+### Proving
+
+Once a node fills a slot, it MUST periodically, yet non-deterministically provide proof to the smart contract that it 
+stores the data it should. Nodes SHOULD detect that a proof is required using the `isProofRequired(slotId)` or that it will
+be required using the `willProofBeRequired(slotId)` in case the node is in [downtime](https://github.com/codex-storage/codex-research/blob/41c4b4409d2092d0a5475aca0f28995034e58d14/design/storage-proof-timing.md).
```

Comment:
I don't think we need to specify periodicity and non-determinicity, as these are defined by the contract. The node only needs to submit proofs to the contract when challenges are issued to the node by the contract.

```suggestion
Once a node fills a slot, it MUST submit proofs to the smart contract when a challenge is issued by the contract.  Nodes SHOULD detect that a proof is required for the current period using the `isProofRequired(slotId)` function,  or that it will
be required using the `willProofBeRequired(slotId)` function in the case that the [pointer is in downtime](https://github.com/codex-storage/codex-research/blob/41c4b4409d2092d0a5475aca0f28995034e58d14/design/storage-proof-timing.md).
```